enhancements/network/openstack-dual-stack.md

```diff
@@ -0,0 +1,175 @@
+---
+title: openstack-dual-stack
+authors:
+  - "@maysamacedo"
+reviewers:
+  - "@mandre for SME review on OpenStack and general feedback"
+  - "@cybertron for SME review on on-premise networking"
+  - "@patrickdillon for SME review and approval on installer"
+approvers:
+  - "@mandre"
+api-approvers:
+  - "None"
+creation-date: 2023-03-13
+last-updated: 2023-03-13
+tracking-link:
+  - https://issues.redhat.com/browse/OCPBU-199
+  - https://issues.redhat.com/browse/OSASINFRA-1938
+---
+
+# OpenStack dual-stack
+
+Covers adding support of OpenShift dual-stack networking on OpenStack.
+
+## Summary
+
+Customers need support for OpenShift clusters running on OpenStack with IPv4 and IPv6.
+
+## Motivation
+
+Customers need dual-stack support to allow migration of workloads from IPv4 to IPv6
+as they are running out of IPv4 addresses. This will allow openshift on openstack to
+serve a wider variety of workloads.
+
+### User Stories
+
+* As an user of a IPI OpenShift cluster, I want to have ingress and egress working with dual-stack.
+* As an user of a IPI OpenShift cluster, I want clusterIP services to be working with dual-stack.
+
+### Goals
+
+- OpenStack VMs are configured with one dual-stack interface.
+- The IPv6 Subnets can be configured with: Stateless, SLAAC or Stateful.
+- The addresses on the VM's interface are configured with the same address
+assigned by OpenStack to the Port.
+- Nodes are configure with both IPv4 and IPv6 addresses, without exposing
+additional addresses, like storage interfaces.
+- Pods are configured with both IPv4 and IPv6 addresses.
+
+### Non-Goals
+
+- Support for OpenStack endpoints configured with dual-stack.
+- Support for LoadBalancer type Services.
+- Installer (IPI) handles creation of the Network and dual-stack Subnet.
+
+## Proposal
+
+Customers will be able to specify the IPv4 and IPv6 Subnets
+meant to be used in the `install-config.yaml`.
+
+### Workflow Description
+
+1. The customers will configure a network (tenant or provider) with IPv4 and
+IPv6 subnets. The IPv6 address mode for the Subnet can be of any type:
+SLAAC, Stateless or Stateful.
+2. The customer will create API and Ingress Ports using the Network created in the
+previous step.
+3. The customer will write the `install-config.yaml` and add the subnet IDs to
+the new field `machinesSubnets`.
+4. The customer will generate the manifests and configured MachineConfigs
```

Comment:
Would there be a way to know which more is enabled? If yes, in baremetal-runtimecfg we could run the detection code (e.g. run some low level commands, to figure out what mode is enabled if that's even possible), then when parsing the templates in MCO, it would populate the right ip_addr_mode. e.g. `--ipv6-mode slaac` passed to runtimecfg command.

@cybertron might have some ideas/opinions on that ^^^

But I think it'll be nice to automate this rather than trusting the customer to do it themselves.

Reply:
@pierreprinetti I would say we can document that, enforcing only one type of ipv6 installation per cluster.

@mkowalski hey o/ perhaps do you have some thoughts if it would be possible to automate the interfaces configuration on MCO?

Reply:
As for the detection, I think this is relatively simple - we can use `ip addr` to detect whether IPv6 address on the interface is stateful or stateless by doing `ip -6 addr show permanent` and `ip -6 addr show dynamic` respectively. As per `man ip-address`

```
dynamic and permanent

(IPv6 only) only list addresses installed due to stateless address
configuration or only list permanent (not dynamic) addresses.
```

As for the configuration with both stateless and stateful addressing, I don't think this is any problem. This is a valid configuration from the linux point of view, so in such a scenario you simply want to apply both stanzas (one for eui64 and the other for kernel params).

Now the question is which component should be responsible for executing this `ip -6 a [...]` command. I could say that it's about generating a correct MachineConfig yaml, so it should be Machine Config Operator. Dispatching this operation to baremetal-runtimecfg somehow feels wrong to me, but I can't justify why, maybe @cybertron can say more (or justify why baremetal-runtimecfg is a good place)

Reply:
Can the eui64 setting not just be the default? If deployers want to change it they can do so on a per-connection basis.

The ip=dhcp,dhcp6 setting is a bit trickier because that does need to be set appropriately for the type of deployment being done. On baremetal that's handled in the installer and baremetal-operator. Is there not something comparable for OpenStack that could be used?

Come to think of it, that could be conditionalized on the IPFamilies template value in MCO. We already do similar things elsewhere. Then it wouldn't have to be set by the user either: https://github.com/openshift/machine-config-operator/blob/60de0ad4be10a7fa84305dc1fe1ad53223d700af/templates/worker/01-worker-kubelet/_base/units/kubelet.service.yaml#L32

That seems like the way to go here.

Reply:
Interesting adding eui64 and method=auto by default worked for all ipv6 addresses mode: slaac/stateful/stateless.
Good suggestion!

https://github.com/openshift/machine-config-operator/pull/3705